Closes #1218 extend pdarray setops to work for multiple arrays

Ethan-DeBandi99 opened the pull request:
This PR (closes #1218):

In order to support set operations for `SegArray` objects efficiently, the set operations for `pdarray` have been updated to support computation on multiple 'flattened' arrays.  The set operations supported are:
- union
- interset
- difference
- symmetric difference

The functionality has been added to the 1d set operations for `pdarray`. The client side code has been updated to detect if the call is being made on 2 `pdarray` objects or 2 iterables (tuple/list) of `pdarray` objects. The iterables should have to objects within: a segments/offsets pdarray and a values pdarray. 

```python
# example of iterables accepted by set operation 1d functions
(segements, values)
# OR
[segments, values]
```

Messages have been added to the server to process the multiple array case.
- `union1d_multi`
- `intersect1d_multi`
- `setdiff1d_multi`
- `setxor1d_multi`
These new processing messages parse out the request message and separate the segments so each can be processed. @pierce314159 mentioned it may be more efficient to utilize `lowLevelLocalizingSlice()`, but this requires other existing utilized functionality to be updated to utilize pointers instead of the pdarray. I included `TODO` tags where this would be utilized if we would like to do it in the future.

In order to configure union1d to function properly, `uniqueFromSorted()` needed to be updated to set the initial truth value at the lowest index on the domain passed instead of always using 0. 

Testing has been added on the client to validate int64 and uint64 functionality as well as error handling.

```python
# Example of flattening multiple arrays
a = [0, 1, 2]
b = [3, 4, 5]
c = [6, 7, 8]
d = [9, 10, 11]
m1_segs = ak.array([0, len(a)])
m1_vals = ak.array(a+b)
m2_segs = ak.array([0, len(c)])
m2_vals = ak.array(c+d)

# Example of union to see how values passed/ returned. Return is always tuple
# Can be passed as tuple
(segments, values) = ak.union1d((m1_segs, m1_vals), (m2_segs, m2_vals))
# OR list
(segments, values) = ak.union1d([m1_segs, m1_vals], [m2_segs, m2_vals])

segments
array([0, 6])

values
array([0, 1, 2, 6, 7, 8, 3, 4, 5, 9, 10, 11])

# Return equates to the arrays
[0, 1, 2, 6, 7, 8]
[3, 4, 5, 9, 10, 11]
```

This PR is required for updates requested in PR #1213.

hokiegeek2 commented on arkouda/pdarraysetops.py:
```diff
@@ -491,25 +572,45 @@ def setxor1d(pda1: pdarray, pda2: pdarray,
-    aux = aux[aux_sort_indices]
-    flag = concatenate((array([True]), aux[1:] != aux[:-1], array([True])))
-    return aux[flag[1:] & flag[:-1]]
+    if isinstance(pda1, pdarray) or isinstance(pda2, pdarray):
```
I recommend some helper methods such as for checking if two params are of the sam type

hokiegeek2 commented on arkouda/pdarraysetops.py:
```diff
@@ -432,30 +488,54 @@ def setdiff1d(pda1: pdarray, pda2: pdarray,
-        pda1 = cast(pdarray, unique(pda1))
-        pda2 = cast(pdarray, unique(pda2))
-    return pda1[in1d(pda1, pda2, invert=True)]
+    if isinstance(pda1, pdarray) or isinstance(pda2, pdarray):
```
as suggested below, recommend helper methods for checking if two params are of the same type, could even add 'em as a decorator

hokiegeek2 commented on src/ArraySetopsMsg.chpl:
```diff
@@ -84,6 +85,143 @@ module ArraySetopsMsg
+          const intx_segs = (+ scan intx_lens) - intx_lens;
+          var intx_vals = makeDistArray((+ reduce intx_lens), int);
+
+          // Compute the union and add values to the corresponding indexes in values
```
I like all the comments in this code -> well done!

hokiegeek2 commented on src/ArraySetopsMsg.chpl:
```diff
@@ -196,6 +471,143 @@ module ArraySetopsMsg
+        size2 = s2_str: int;
+      }
+      catch {
+        var errorMsg = "Count could not be interpretted as an int1: %s, int2: %s)".format(s1_str, s2_str);
```
typo "interpretted"

Ethan-DeBandi99 replied:
Corrected. Thanks!

hokiegeek2 reviewed (approved):
Overrall looks great to me (especially all the comments!), just a few recommendations and one type fix

bmcdonald3 commented on arkouda/pdarraysetops.py:
```diff
@@ -309,47 +311,76 @@ def union1d(pda1: pdarray, pda2: pdarray) -> pdarray:
-    return cast(pdarray,
-                unique(cast(pdarray,
-                            concatenate((unique(pda1), unique(pda2)), ordered=False))))  # type: ignore
+    if isinstance(pda1, pdarray) or isinstance(pda2, pdarray):
+        if not (isinstance(pda1, pdarray) and isinstance(pda2, pdarray)):
```
Could this check be changed to just a single `if isinstance(pda1, pdarray) and isinstance(pda2, pdarray)`? It seems equivalent to me, but I'm not certain

bmcdonald3 replied:
I see, thanks for clarifying.

bmcdonald3 reviewed (approved):
This looks good to me, just one minor comment. I also was slightly confused by the `segs` name because it kept making me think of the segmented (strings) array in Arkouda, but I don't have a problem with it.

Ethan-DeBandi99 commented on arkouda/pdarraysetops.py:
```diff
@@ -309,47 +311,76 @@ def union1d(pda1: pdarray, pda2: pdarray) -> pdarray:
-    return cast(pdarray,
-                unique(cast(pdarray,
-                            concatenate((unique(pda1), unique(pda2)), ordered=False))))  # type: ignore
+    if isinstance(pda1, pdarray) or isinstance(pda2, pdarray):
```
I had this configured because the inputs need to be either both pdarrays or both iterable (tuple/list). Checking that 1 is a pdarray, then allows for the 2nd check to error if both are not. @pierce314159 and I looked at this a bit and thought this was the most efficient way to get the desired result with the type checking. I'd be happy to simplify it if you have a suggestion on how, but this is the only way I am seeing to verify that if one input is a pdarray, both are without have to do 
```python
if isinstance(pda1, pdarray) and isinstance(pda2, pdarray):
    code to process pdarrays
elif (isinstance(pda1, list) or isinstance(pda1, tuple)) and (isinstance(pda2, list) or isinstance(pda2, tuple)):
    code to process iterables
else:
    raise TypeError
```

bmcdonald3 commented:
One of my PRs is having this same issue as this one with the CI where it just says `Workflow runs completed with no jobs` and isn't actually running the CI... I don't see any changes to the CI code or anything, I'm not sure what is going on. Does anyone else have any ideas?

Ethan-DeBandi99 commented:
> One of my PRs is having this same issue as this one with the CI where it just says `Workflow runs completed with no jobs` and isn't actually running the CI... I don't see any changes to the CI code or anything, I'm not sure what is going on. Does anyone else have any ideas?

I was having this issue a lot earlier in the week. It looks like it just takes a while for the jobs to kick off. Earlier this week they were showing as queued, but right now I am not seeing anything. If things are delayed, based on what happened earlier this week, it should come back up soon.

pierce314159 commented on tests/setops_test.py:
```diff
@@ -92,7 +92,37 @@ def testSetxor1d(self):
+        self.assertListEqual(segs.to_ndarray().tolist(), [0, 4])
+        self.assertListEqual(vals.to_ndarray().tolist(), [1, 4, 5, 7, 1, 3, 9])
+
+        segs, vals = ak.setxor1d((ak.array([0, len(a)]), ak.array(a + b, dtype=ak.int64)),
+                                  (ak.array([0, len(c)]), ak.array(c + d, dtype=ak.int64)))
```
should this be `ak.uint64`? Same question for the other tests

Ethan-DeBandi99 replied:
You are correct. I updated this to `ak.uint64`.

pierce314159 reviewed (commented):
I think this looks great! Couple in line comments to check out

One suggestion to reduce duplicated chapel logic between the ops. I think it's worthwhile to collapse these into one chapel func to avoid having to remember to change multiple spots for any small updates. something like

```python
def intersect1d():
  subcmd = 'intersect'
  repMsg = cast(str, generic_msg(cmd="setop1d_multi", 
                                 args=f"{subcmd} {pda1[0].name} {pda1[1].name} {pda1[1].size} {pda2[0].name} 
```

```chapel
proc setop1d_multiMsg   {
  // do preprocessing on segs, len calc, mem check
  select subcmd {
    when "intersect" {
      // double intersect forall
    }
    when "union" {
      // double union forall
    }
    ...
  }
}
```
I think this will be easier to maintain with ~150 fewer lines of boilerplate code but im open to change my mind on this

Ethan-DeBandi99 commented on arkouda/pdarraysetops.py:
```diff
@@ -275,7 +276,8 @@ def concatenate(arrays: Sequence[Union[pdarray, Strings, 'Categorical']],  # typ
 # (A1 | A2) Set Union: elements are in one or the other or both
 @typechecked
-def union1d(pda1: pdarray, pda2: pdarray) -> pdarray:
+def union1d(pda1: Union[pdarray, List[pdarray], tuple], pda2: Union[pdarray,
+            List[pdarray], tuple]) -> Union[pdarray, tuple, str]:
```
During my testing having it configured as `Tuple[pdarray]` caused an error. Maybe the definition is slightly different.I will double check to make sure I didn't put something in wrong.

Ethan-DeBandi99 replied:
I updated this to `Tuple[pdarray, pdarray]`. I initially forget the 2nd `pdarray` which was causing it to fail.

Ethan-DeBandi99 commented:
@pierce314159 Adding a singular comment to address everything, since it will take me a bit to update. I agree with all your suggestions. The only other way I can think to simplify the boilerplate code in the functions is to add a helper. I think your proposed solution is cleaner, so I am going to reformat the code to function in that manor.

Ethan-DeBandi99 commented on src/ArraySetopsMsg.chpl:
```diff
@@ -84,6 +85,143 @@ module ArraySetopsMsg
+        }
+        if (i == low || m2 < l2) {
+          m2 = l2;
+        }
```
Collapsed functions and updated max calculation to use the reduce.

Ethan-DeBandi99 commented on src/ArraySetopsMsg.chpl:
```diff
@@ -84,6 +85,143 @@ module ArraySetopsMsg
+        size2 = s2_str: int;
+      }
+      catch {
+        var errorMsg = "Count could not be interpreted as an int1: %s, int2: %s)".format(s1_str, s2_str);
```
Updated to suggested error message.

Ethan-DeBandi99 commented:
> I think this looks great! Couple in line comments to check out
> 
> One suggestion to reduce duplicated chapel logic between the ops. I think it's worthwhile to collapse these into one chapel func to avoid having to remember to change multiple spots for any small updates. something like
> 
> ```python
> def intersect1d():
>   subcmd = 'intersect'
>   repMsg = cast(str, generic_msg(cmd="setop1d_multi", 
>                                  args=f"{subcmd} {pda1[0].name} {pda1[1].name} {pda1[1].size} {pda2[0].name} 
> ```
> 
> ```chapel
> proc setop1d_multiMsg   {
>   // do preprocessing on segs, len calc, mem check
>   select subcmd {
>     when "intersect" {
>       // double intersect forall
>     }
>     when "union" {
>       // double union forall
>     }
>     ...
>   }
> }
> ```
> 
> I think this will be easier to maintain with ~150 fewer lines of boilerplate code but im open to change my mind on this

I updated this to use 1 message with subcommands. I also moved the individual functionality into `ArraySetops.chpl` to make reading the `setops1d_multiMsg()` function easier to read and less bloated.

pierce314159 reviewed (changes requested):
The majority of the code looks great! But i might have discovered a potential remote-to-remote aggregation issue, im going to mark this as request changes until I'm sure. If @ronawho could advise on it, that'd be great

I'm also nitpicking about comments and some docstring examples cause that's the way i am lol

Ethan-DeBandi99 commented on src/ArraySetops.chpl:
```diff
@@ -39,6 +42,31 @@ module ArraySetops
+    proc intersect1d_multi(segments1: borrowed SymEntry(int), values1: borrowed SymEntry(?t), lens1: [] int, segments2: borrowed SymEntry(int), values2: borrowed SymEntry(t), lens2: [] int, isUnique: bool) throws {
+      var intx_lens: [segments1.aD] int;
+
+      // Compute lengths of the segments resulting from each union
```
Corrected comments to match functions they are in

Ethan-DeBandi99 commented on src/ArraySetops.chpl:
```diff
@@ -39,6 +42,31 @@ module ArraySetops
+        var intx = intersect1d(values1.a[s1..#l1], values2.a[s2..#l2], isUnique);
+        for i in (0..#il){
+          agg.copy(intx_vals[i+is], intx[i]);
+        }
```
Thanks for catching this. I didn't even think about it. I made this change in each setop function.

ronawho replied:
I don't fully understand what this code is trying to do, but at a glance `values1.a[s1..#l1]` should be creating a local array, and I don't think `intersect1d` would do anything to then distributed a local array, but I haven't checked. So in the current code I _think_ `intx` should always be local.

That said, I don't think this code will be fast because it will take time to create those local arrays, when in the common case you're probably just operating on a local portion of the distributed array. This is was `lowLevelLocalizingSlice` was originally added for -- to make it cheaper to operate on portions of likely local, but possibly remote arrays.

However, for sorted strings Bill was seeing performance for bringing the data to the current node was quite poor because it was no longer the common case that data was local, so he introduced the `computeOnSegments` code as a way to always operate on local portions of strings. https://github.com/Bears-R-Us/arkouda/issues/1101 has more information on that, and I wonder if extending `computeOnSegments` is what you'd want to be doing here.

Ethan-DeBandi99 replied:
I believe you may be correct for intersect. The other functions are all using the uniqueSort, which returns a new distributed array. I just lumped everything together. I can update that.

This code is definitely not going to be the best performance wise. The intention was to move these operations to the server because it will still perform better than handling everything client side. I have not yet, but will be adding an issue to address the performance of these set operations.

pierce314159 replied:
I agree with Ethan, intersect is calling uniquesort which is returning a distributed array

ronawho replied:
Hmm, I thought most of the sort routines had the output type match the input type, but I could easily be wrong here (and I haven't looked at the code yet, so I trust what you're saying over my guess)

Ethan-DeBandi99 commented on src/ArraySetopsMsg.chpl:
```diff
@@ -251,6 +252,165 @@ module ArraySetopsMsg
+
+              repMsg = "created " + st.attrib(s_name) + "+created " + st.attrib(v_name);
+              asLogger.debug(getModuleName(),getRoutineName(),getLineNumber(),repMsg);
+              return new MsgTuple(repMsg, MsgType.NORMAL);
```
Moved the block and removed the repeated code

pierce314159 reviewed (approved):
I have no issues, great job!

mhmerrill commented:
@bmcdonald3 @pierce314159 @Ethan-DeBandi99 @hokiegeek2 did you guys get all your conversations resolved ??? is this stuff ready to be merged? Did @ronawho weight in on the use of the aggregator in the one conversation left unresolved?

hokiegeek2 commented:
I approved the merge request, all good from my perspective
